src/algorithms/calorimetry/CalorimeterClusterRecoCoG.cc

```diff
@@ -161,3 +162,162 @@ void CalorimeterClusterRecoCoG::AlgorithmProcess() {
+    if (eta < minHitEta) {
+      minHitEta = eta;
+    }
+    if (eta > maxHitEta) {
+      maxHitEta = eta;
+    }
+  }
+  cl.setEnergy(totalE / m_sampFrac);
+  cl.setEnergyError(0.);
+  cl.setTime(time);
+  cl.setTimeError(timeError);
+
+  // center of gravity with logarithmic weighting
+  float tw = 0.;
+  auto v   = cl.getPosition();
+  for (unsigned i = 0; i < pcl->getHits().size(); ++i) {
+    const auto& hit   = pcl->getHits()[i];
+    const auto weight = pcl->getWeights()[i];
+    //      _DBG_<<" -- weight = " << weight << "  E=" << hit.getEnergy() << " totalE=" <<totalE << " log(E/totalE)=" << std::log(hit.getEnergy()/totalE) << std::endl;
+    float w           = weightFunc(hit.getEnergy() * weight, totalE, m_logWeightBase, 0);
+    tw += w;
+    v = v + (hit.getPosition() * w);
+  }
+  if (tw == 0.) {
+    m_log->warn("zero total weights encountered, you may want to adjust your weighting parameter.");
+  }
+  cl.setPosition(v / tw);
+  cl.setPositionError({}); // @TODO: Covariance matrix
+
+  // Optionally constrain the cluster to the hit eta values
+  if (m_enableEtaBounds) {
+    const bool overflow  = (edm4eic::eta(cl.getPosition()) > maxHitEta);
+    const bool underflow = (edm4eic::eta(cl.getPosition()) < minHitEta);
+    if (overflow || underflow) {
+      const double newEta   = overflow ? maxHitEta : minHitEta;
+      const double newTheta = edm4eic::etaToAngle(newEta);
+      const double newR     = edm4eic::magnitude(cl.getPosition());
+      const double newPhi   = edm4eic::angleAzimuthal(cl.getPosition());
+      cl.setPosition(edm4eic::sphericalToVector(newR, newTheta, newPhi));
+      m_log->debug("Bound cluster position to contributing hits due to {}", (overflow ? "overflow" : "underflow"));
+    }
+  }
+
+  // Additional convenience variables
+
+  // best estimate on the cluster direction is the cluster position
+  // for simple 2D CoG clustering
+  cl.setIntrinsicTheta(edm4eic::anglePolar(cl.getPosition()));
+  cl.setIntrinsicPhi(edm4eic::angleAzimuthal(cl.getPosition()));
+  // TODO errors
+
+  //_______________________________________
+  // Calculate cluster profile:
+  //    radius,
+  //   	dispersion (energy weighted radius),
+  //   	sigma_long
+  //   	sigma_short
+  //   	sigma_z
+  double radius = 0, dispersion = 0, lambda_1 = 0, lambda_2 = 0, lambda_3 = 0;
+  double w_sum = 0;
+  double sum_11 = 0, sum_22 = 0, sum_33 = 0;
+  double sum_12 = 0, sum_13 = 0, sum_23 = 0;
+  double sum_1 = 0, sum_2 = 0, sum_3 = 0;
+
+  if (cl.getNhits() > 1) {
+
+    for (const auto& hit : pcl->getHits()) {
+
+      const auto delta = cl.getPosition() - hit.getPosition();
+      radius += delta * delta;
+
+      double w = weightFunc(hit.getEnergy(), cl.getEnergy(), m_logWeightBase, 0);
+      w_sum += w;
+      dispersion += delta * delta * w;
+
+      double pos_1 = edm4eic::anglePolar( hit.getPosition() );
+      double pos_2 = edm4eic::angleAzimuthal( hit.getPosition() );
+      double pos_3 = hit.getPosition().z;
```

Comment:
I don't understand how this works to put these into a matrix. From the point of dimensional analysis it would make sense to do:
```suggestion
      double pos_1 = edm4eic::anglePolar( hit.getPosition() ) * hit.getPosition().mag();
      double pos_2 = edm4eic::angleAzimuthal( hit.getPosition() ) * hit.getPosition().mag();
      double pos_3 = hit.getPosition().z;
```
But then, we might as well
```suggestion
      double pos_1 = hit.getPosition().x;
      double pos_2 = hit.getPosition().y;
      double pos_3 = hit.getPosition().z;
```
which is what `m_xyClusterProfiling` does. Perhaps, we only need one calculation variant always?

Reply:
Yes, it's a bit odd to mix eta-phi with z, but I think the scale (and units) of each dimension may not really matter.  x-y-z dimensions is much more intuitive, I agree.  I included the eta-phi-z because other collaborations like ALICE have used eta-phi in their 2D analyses.  We can exclude the eta-phi-z for now if needed.  What do you think?

Reply:
x-y-z makes sense and theta-phi or eta-phi makes sense, but not theta-phi-z. The dimension matters. For an argument, imagine, we present your algorithm not with mm (default length unit in EDM4hep), but in, say, meters or in micrometers. That will change directions of your eigenvectors. Also note that the z and the polar basis vectors are collinear at eta=0.

Reply:
I now have both 2D (theta-phi) and 3D (x-y-z) shape analyses included separately.  All 5=2+3 eigenvalues are included in the shape parameter list.